File: scripts/parse_judgment.py
"""Parse and validate raw model judgment output into judgment_patches.json."""

from __future__ import annotations

import json
import math
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

if hasattr(sys.stdin, 'reconfigure'):
    sys.stdin.reconfigure(encoding='utf-8')
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')
if hasattr(sys.stderr, 'reconfigure'):
    sys.stderr.reconfigure(encoding='utf-8')

# ---------------------------------------------------------------------------
# Level-range tables (canonical from rubric docs)
# ---------------------------------------------------------------------------

# Each entry: (level, lower_bound, upper_bound)
# Sorted highest-level-first for scan resolution
F3_LEVELS: list[tuple[int, float, float]] = [
    (4, 0.90, 1.00),
    (3, 0.65, 0.85),
    (2, 0.40, 0.60),
    (1, 0.15, 0.35),
    (0, 0.00, 0.10),
]

F8_LEVELS: list[tuple[int, float, float]] = [
    (3, 0.85, 1.00),
    (2, 0.55, 0.80),
    (1, 0.30, 0.50),
    (0, 0.10, 0.25),
]

REASONING_MAX_LEN = 80
KNOWN_PATCH_FIELDS = {"F6_patch", "F7_patch", "F1_patch"}
JUDGMENT_FACTORS = {"F3", "F8"}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _warn(msg: str) -> None:
    print(f"WARNING: {msg}", file=sys.stderr)


def _fatal(msg: str) -> None:
    print(f"FATAL: {msg}", file=sys.stderr)
    sys.exit(1)


_OPENING_FENCE_PATTERN = re.compile(r'^```\w*\s*\n')
_CLOSING_FENCE_PATTERN = re.compile(r'\n```\s*$')


def strip_fences(text: str) -> str:
    """Remove markdown code fences wrapping the JSON."""
    text = text.strip()
    text = _OPENING_FENCE_PATTERN.sub('', text)
    text = _CLOSING_FENCE_PATTERN.sub('', text)
    return text


def _find_balanced_array(text: str, start: int) -> int | None:
    """Find the matching ']' for text[start]=='[' respecting string literals."""
    depth = 0
    in_string = False
    escape = False
    for i in range(start, len(text)):
        ch = text[i]
        if escape:
            escape = False
            continue
        if ch == '\\' and in_string:
            escape = True
            continue
        if ch == '"' and not escape:
            in_string = not in_string
            continue
        if in_string:
            continue
        if ch == '[':
            depth += 1
        elif ch == ']':
            depth -= 1
            if depth == 0:
                return i
    return None


def extract_json_array(text: str) -> list[dict]:
    """Find the outermost JSON array in text using bracket-balance counting."""
    pos = 0
    while True:
        start = text.find('[', pos)
        if start == -1:
            break

        end = _find_balanced_array(text, start)
        if end is None:
            pos = start + 1
            continue

        substr = text[start:end + 1]
        try:
            parsed = json.loads(substr)
        except json.JSONDecodeError:
            pos = start + 1
            continue

        if isinstance(parsed, list):
            return parsed

        pos = start + 1

    raise ValueError("No valid JSON array found in input")


def resolve_level(value: float, level_table: list[tuple[int, float, float]]) -> int:
    """Resolve a value to the highest level where value >= lower_bound."""
    for level, lower, upper in level_table:
        if value >= lower:
            return level
    return level_table[-1][0]


def level_midpoint(level: int, level_table: list[tuple[int, float, float]]) -> float:
    """Return the midpoint of a level's range."""
    for lev, lower, upper in level_table:
        if lev == level:
            return round((lower + upper) / 2, 2)
    return 0.50


def validate_value_range(
    rule_id: str,
    factor_name: str,
    value: float | None,
    stated_level: int | None,
    level_table: list[tuple[int, float, float]],
) -> tuple[float | None, int | None]:
    """Validate that value falls within the stated level's range; level wins on mismatch."""
    if value is None or stated_level is None:
        return (None, None)

    stated_range = None
    for lev, lower, upper in level_table:
        if lev == stated_level:
            stated_range = (lower, upper)
            break

    if stated_range is None:
        _warn(f"{rule_id} {factor_name}: unknown level {stated_level}, keeping value {value}")
        return (value, stated_level)

    lower, upper = stated_range
    if lower <= value <= upper:
        return (value, stated_level)

    resolved_level = resolve_level(value, level_table)
    if resolved_level == stated_level:
        return (value, stated_level)

    corrected_value = level_midpoint(stated_level, level_table)
    _warn(
        f"{rule_id} {factor_name}: value {value} outside level {stated_level} "
        f"range [{lower}, {upper}]. Corrected to {corrected_value} (level midpoint)."
    )
    return (corrected_value, stated_level)


def truncate_reasoning(reasoning: str) -> str:
    """Truncate reasoning to REASONING_MAX_LEN characters."""
    if len(reasoning) <= REASONING_MAX_LEN:
        return reasoning
    return reasoning[:REASONING_MAX_LEN - 3] + "..."


def validate_factor(rule_id: str, factor_name: str, factor_data: dict, level_table: list[tuple[int, float, float]]) -> dict | None:
    """Validate and normalize a single factor entry; returns None if entirely malformed."""
    if not isinstance(factor_data, dict):
        _warn(f"{rule_id} {factor_name}: expected dict, got {type(factor_data).__name__}")
        return None

    value = factor_data.get("value")
    level = factor_data.get("level")
    reasoning = factor_data.get("reasoning", "")

    if value is None and level is None:
        return {
            "value": None,
            "level": None,
            "reasoning": truncate_reasoning(str(reasoning)) if reasoning else "model_could_not_score",
        }

    if not isinstance(value, (int, float)):
        _warn(f"{rule_id} {factor_name}: value must be a number, got {type(value).__name__}")
        return None
    if not isinstance(level, int):
        _warn(f"{rule_id} {factor_name}: level must be an integer, got {type(level).__name__}")
        return None

    value = float(value)

    if value < 0.0 or value > 1.0:
        _warn(f"{rule_id} {factor_name}: value {value} outside [0, 1], clamping")
        value = max(0.0, min(1.0, value))

    value, level = validate_value_range(rule_id, factor_name, value, level, level_table)

    return {
        "value": value,
        "level": level,
        "reasoning": truncate_reasoning(str(reasoning)),
    }


def validate_entry(entry: dict, idx: int) -> tuple[str | None, dict]:
    """Validate a single array entry; returns (rule_id, patches) or (None, {}) on failure."""
    if not isinstance(entry, dict):
        _warn(f"Entry {idx}: expected dict, got {type(entry).__name__}, skipping")
        return (None, {})

    rule_id = entry.get("id")
    if not rule_id or not isinstance(rule_id, str):
        _warn(f"Entry {idx}: missing or invalid 'id' field, skipping")
        return (None, {})

    patches = {}

    f3_data = entry.get("F3")
    if f3_data is not None:
        validated = validate_factor(rule_id, "F3", f3_data, F3_LEVELS)
        if validated is not None:
            patches["F3"] = validated
        else:
            patches["F3"] = {"value": None, "level": None, "reasoning": "parse_validation_failed"}
    else:
        patches["F3"] = {"value": None, "level": None, "reasoning": "model_omitted"}
        _warn(f"{rule_id}: F3 missing from entry")

    f8_data = entry.get("F8")
    if f8_data is not None:
        validated = validate_factor(rule_id, "F8", f8_data, F8_LEVELS)
        if validated is not None:
            patches["F8"] = validated
        else:
            patches["F8"] = {"value": None, "level": None, "reasoning": "parse_validation_failed"}
    else:
        patches["F8"] = {"value": None, "level": None, "reasoning": "model_omitted"}
        _warn(f"{rule_id}: F8 missing from entry")

    for field_name in KNOWN_PATCH_FIELDS:
        if field_name in entry:
            patch_data = entry[field_name]
            if not isinstance(patch_data, dict):
                _warn(f"{rule_id}: {field_name} must be an object with a 'value' key, got {type(patch_data).__name__}; dropping")
                continue
            if "value" not in patch_data:
                _warn(f"{rule_id}: {field_name} is missing required 'value' key; dropping (keys present: {sorted(patch_data.keys())})")
                continue
            val = patch_data["value"]
            if val is not None and not isinstance(val, (int, float)):
                _warn(f"{rule_id}: {field_name}.value must be a number or null, got {type(val).__name__}; dropping")
                continue
            if isinstance(val, (int, float)) and not (0.0 <= float(val) <= 1.0):
                _warn(f"{rule_id}: {field_name}.value must be in [0, 1], got {val}; dropping")
                continue
            if "reasoning" in patch_data:
                patch_data["reasoning"] = truncate_reasoning(str(patch_data["reasoning"]))
            patches[field_name] = patch_data

    return (rule_id, patches)


def build_patches(entries: list[dict], expected_ids: set[str]) -> dict:
    """Transform validated array entries into patches map; handle missing IDs."""
    patches = {}
    seen_ids = set()

    for idx, entry in enumerate(entries):
        rule_id, rule_patches = validate_entry(entry, idx)
        if rule_id is None:
            continue
        if rule_id in seen_ids:
            _warn(f"{rule_id}: duplicate entry, last one wins")
        seen_ids.add(rule_id)
        patches[rule_id] = rule_patches

    missing_ids = expected_ids - seen_ids
    if missing_ids:
        total = len(expected_ids)
        tolerance = max(2, math.ceil(0.05 * total))

        for mid in sorted(missing_ids):
            _warn(f"{mid}: not found in model output, inserting null entry")
            patches[mid] = {
                "F3": {"value": None, "level": None, "reasoning": "model_omitted"},
                "F8": {"value": None, "level": None, "reasoning": "model_omitted"},
            }

        if len(missing_ids) > tolerance:
            _fatal(
                f"{len(missing_ids)} of {total} rule IDs missing from model output "
                f"(tolerance: {tolerance}). Model may have truncated. "
                f"Missing: {', '.join(sorted(missing_ids)[:10])}{'...' if len(missing_ids) > 10 else ''}"
            )

    unexpected = seen_ids - expected_ids
    for uid in sorted(unexpected):
        _warn(f"{uid}: in model output but not in scored_semi.json, ignoring")
        patches.pop(uid, None)

    return patches


def load_expected_rule_ids(scored_semi_path: str) -> set[str]:
    """Load expected rule IDs from scored_semi.json; validates schema_version."""
    with open(scored_semi_path, encoding="utf-8") as f:
        data = json.load(f)

    schema_version = data.get("schema_version")
    if schema_version is None:
        _warn("scored_semi.json has no schema_version field; assuming 0.1")
    elif schema_version != "0.1":
        _fatal(f"scored_semi.json schema_version is '{schema_version}', expected '0.1'. "
               f"Pipeline version skew — regenerate scored_semi.json with the current scripts.")

    rules = data.get("rules", [])
    return {r["id"] for r in rules if "id" in r}


def main() -> None:
    if len(sys.argv) < 2:
        _fatal("Usage: parse_judgment.py <scored_semi.json> [--expected-ids R001,...] "
               "[--input file] [--output file]")

    input_path = None
    output_path = None
    expected_ids_override = None
    positional = []
    args = sys.argv[1:]
    i = 0
    while i < len(args):
        if args[i] == "--input" and i + 1 < len(args):
            input_path = args[i + 1]
            i += 2
        elif args[i] == "--output" and i + 1 < len(args):
            output_path = args[i + 1]
            i += 2
        elif args[i] == "--expected-ids" and i + 1 < len(args):
            expected_ids_override = set(args[i + 1].split(","))
            i += 2
        else:
            positional.append(args[i])
            i += 1

    if not positional:
        _fatal("Missing scored_semi.json argument")
    scored_semi_path = positional[0]

    try:
        all_ids = load_expected_rule_ids(scored_semi_path)
    except (FileNotFoundError, json.JSONDecodeError, KeyError) as e:
        _fatal(f"Cannot read scored_semi.json: {e}")

    if not all_ids:
        _fatal("scored_semi.json contains no rules")

    expected_ids = expected_ids_override if expected_ids_override else all_ids

    if input_path:
        with open(input_path, encoding="utf-8") as f:
            raw_input = f.read()
    else:
        raw_input = sys.stdin.read()
    if not raw_input.strip():
        _fatal("Empty input" + (f" from {input_path}" if input_path else " on stdin"))

    cleaned = strip_fences(raw_input)

    try:
        entries = extract_json_array(cleaned)
    except ValueError as e:
        _fatal(str(e))

    patches = build_patches(entries, expected_ids)

    output = {
        "schema_version": "0.1",
        "model_version": "unknown",
        "patches": patches,
    }
    if output_path:
        with open(output_path, "w", encoding="utf-8") as f:
            json.dump(output, f, indent=2, ensure_ascii=False)
    else:
        json.dump(output, sys.stdout, indent=2, ensure_ascii=False)
        sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: scripts/parse_judgment.js
"use strict";

const fs = require("fs");

function warn(msg) {
  process.stderr.write(`WARNING: ${msg}\n`);
}

function fatal(msg) {
  process.stderr.write(`FATAL: ${msg}\n`);
  process.exit(1);
}

const F3_LEVELS = [
  [4, 0.9, 1.0],
  [3, 0.65, 0.85],
  [2, 0.4, 0.6],
  [1, 0.15, 0.35],
  [0, 0.0, 0.1],
];

const F8_LEVELS = [
  [3, 0.85, 1.0],
  [2, 0.55, 0.8],
  [1, 0.3, 0.5],
  [0, 0.1, 0.25],
];

const REASONING_MAX_LEN = 80;
const KNOWN_PATCH_FIELDS = ["F6_patch", "F7_patch", "F1_patch"];

const OPENING_FENCE_PATTERN = /^```\w*\s*\n/;
const CLOSING_FENCE_PATTERN = /\n```\s*$/;

function stripFences(text) {
  text = text.trim();
  text = text.replace(OPENING_FENCE_PATTERN, "");
  text = text.replace(CLOSING_FENCE_PATTERN, "");
  return text;
}

function findBalancedArray(text, start) {
  let depth = 0;
  let inString = false;
  let escape = false;
  for (let i = start; i < text.length; i++) {
    const ch = text[i];
    if (escape) {
      escape = false;
      continue;
    }
    if (ch === "\\" && inString) {
      escape = true;
      continue;
    }
    if (ch === '"' && !escape) {
      inString = !inString;
      continue;
    }
    if (inString) continue;
    if (ch === "[") {
      depth += 1;
    } else if (ch === "]") {
      depth -= 1;
      if (depth === 0) return i;
    }
  }
  return null;
}

function extractJsonArray(text) {
  let pos = 0;
  while (true) {
    const start = text.indexOf("[", pos);
    if (start === -1) break;

    const end = findBalancedArray(text, start);
    if (end === null) {
      pos = start + 1;
      continue;
    }

    const substr = text.slice(start, end + 1);
    let parsed;
    try {
      parsed = JSON.parse(substr);
    } catch {
      pos = start + 1;
      continue;
    }

    if (Array.isArray(parsed)) return parsed;

    pos = start + 1;
  }

  throw new Error("No valid JSON array found in input");
}

function resolveLevel(value, levelTable) {
  for (const [level, lower] of levelTable) {
    if (value >= lower) return level;
  }
  return levelTable[levelTable.length - 1][0];
}

function levelMidpoint(level, levelTable) {
  for (const [lev, lower, upper] of levelTable) {
    if (lev === level) return Math.round(((lower + upper) / 2) * 100) / 100;
  }
  return 0.5;
}

function validateValueRange(ruleId, factorName, value, statedLevel, levelTable) {
  if (value === null || value === undefined || statedLevel === null || statedLevel === undefined) {
    return [null, null];
  }

  let statedRange = null;
  for (const [lev, lower, upper] of levelTable) {
    if (lev === statedLevel) {
      statedRange = [lower, upper];
      break;
    }
  }

  if (statedRange === null) {
    warn(`${ruleId} ${factorName}: unknown level ${statedLevel}, keeping value ${value}`);
    return [value, statedLevel];
  }

  const [lower, upper] = statedRange;
  if (lower <= value && value <= upper) {
    return [value, statedLevel];
  }

  const resolvedLevel = resolveLevel(value, levelTable);
  if (resolvedLevel === statedLevel) {
    return [value, statedLevel];
  }

  const correctedValue = levelMidpoint(statedLevel, levelTable);
  warn(
    `${ruleId} ${factorName}: value ${value} outside level ${statedLevel} ` +
      `range [${lower}, ${upper}]. Corrected to ${correctedValue} (level midpoint).`
  );
  return [correctedValue, statedLevel];
}

function truncateReasoning(reasoning) {
  if (reasoning.length <= REASONING_MAX_LEN) return reasoning;
  return reasoning.slice(0, REASONING_MAX_LEN - 3) + "...";
}

function pyStr(v) {
  if (v === null || v === undefined) return "None";
  if (v === true) return "True";
  if (v === false) return "False";
  return String(v);
}

function typeName(v) {
  if (v === null || v === undefined) return "NoneType";
  if (Array.isArray(v)) return "list";
  if (typeof v === "boolean") return "bool";
  if (typeof v === "number") return Number.isInteger(v) ? "int" : "float";
  if (typeof v === "string") return "str";
  if (typeof v === "object") return "dict";
  return typeof v;
}

function isDict(v) {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function isNumber(v) {
  return typeof v === "number";
}

function isInt(v) {
  return typeof v === "number" && Number.isInteger(v);
}

function validateFactor(ruleId, factorName, factorData, levelTable) {
  if (!isDict(factorData)) {
    warn(`${ruleId} ${factorName}: expected dict, got ${typeName(factorData)}`);
    return null;
  }

  let value = Object.prototype.hasOwnProperty.call(factorData, "value") ? factorData.value : undefined;
  let level = Object.prototype.hasOwnProperty.call(factorData, "level") ? factorData.level : undefined;
  const reasoning = Object.prototype.hasOwnProperty.call(factorData, "reasoning") ? factorData.reasoning : "";

  if ((value === null || value === undefined) && (level === null || level === undefined)) {
    return {
      value: null,
      level: null,
      reasoning: reasoning ? truncateReasoning(pyStr(reasoning)) : "model_could_not_score",
    };
  }

  if (!isNumber(value)) {
    warn(`${ruleId} ${factorName}: value must be a number, got ${typeName(value)}`);
    return null;
  }
  if (!isInt(level)) {
    warn(`${ruleId} ${factorName}: level must be an integer, got ${typeName(level)}`);
    return null;
  }

  value = Number(value);

  if (value < 0.0 || value > 1.0) {
    warn(`${ruleId} ${factorName}: value ${value} outside [0, 1], clamping`);
    value = Math.max(0.0, Math.min(1.0, value));
  }

  [value, level] = validateValueRange(ruleId, factorName, value, level, levelTable);

  return {
    value,
    level,
    reasoning: truncateReasoning(pyStr(reasoning)),
  };
}

function validateEntry(entry, idx) {
  if (!isDict(entry)) {
    warn(`Entry ${idx}: expected dict, got ${typeName(entry)}, skipping`);
    return [null, {}];
  }

  const ruleId = entry.id;
  if (!ruleId || typeof ruleId !== "string") {
    warn(`Entry ${idx}: missing or invalid 'id' field, skipping`);
    return [null, {}];
  }

  const patches = {};

  const f3Data = entry.F3;
  if (f3Data !== null && f3Data !== undefined) {
    const validated = validateFactor(ruleId, "F3", f3Data, F3_LEVELS);
    if (validated !== null) {
      patches.F3 = validated;
    } else {
      patches.F3 = { value: null, level: null, reasoning: "parse_validation_failed" };
    }
  } else {
    patches.F3 = { value: null, level: null, reasoning: "model_omitted" };
    warn(`${ruleId}: F3 missing from entry`);
  }

  const f8Data = entry.F8;
  if (f8Data !== null && f8Data !== undefined) {
    const validated = validateFactor(ruleId, "F8", f8Data, F8_LEVELS);
    if (validated !== null) {
      patches.F8 = validated;
    } else {
      patches.F8 = { value: null, level: null, reasoning: "parse_validation_failed" };
    }
  } else {
    patches.F8 = { value: null, level: null, reasoning: "model_omitted" };
    warn(`${ruleId}: F8 missing from entry`);
  }

  for (const fieldName of KNOWN_PATCH_FIELDS) {
    if (Object.prototype.hasOwnProperty.call(entry, fieldName)) {
      const patchData = entry[fieldName];
      if (!isDict(patchData)) {
        warn(`${ruleId}: ${fieldName} must be an object with a 'value' key, got ${typeName(patchData)}; dropping`);
        continue;
      }
      if (!Object.prototype.hasOwnProperty.call(patchData, "value")) {
        warn(`${ruleId}: ${fieldName} is missing required 'value' key; dropping (keys present: ${JSON.stringify(Object.keys(patchData).sort())})`);
        continue;
      }
      const val = patchData.value;
      if (val !== null && val !== undefined && !isNumber(val)) {
        warn(`${ruleId}: ${fieldName}.value must be a number or null, got ${typeName(val)}; dropping`);
        continue;
      }
      if (isNumber(val) && !(0.0 <= Number(val) && Number(val) <= 1.0)) {
        warn(`${ruleId}: ${fieldName}.value must be in [0, 1], got ${val}; dropping`);
        continue;
      }
      if (Object.prototype.hasOwnProperty.call(patchData, "reasoning")) {
        patchData.reasoning = truncateReasoning(pyStr(patchData.reasoning));
      }
      patches[fieldName] = patchData;
    }
  }

  return [ruleId, patches];
}

function buildPatches(entries, expectedIds) {
  const patches = {};
  const seenIds = new Set();

  entries.forEach((entry, idx) => {
    const [ruleId, rulePatches] = validateEntry(entry, idx);
    if (ruleId === null) return;
    if (seenIds.has(ruleId)) {
      warn(`${ruleId}: duplicate entry, last one wins`);
    }
    seenIds.add(ruleId);
    patches[ruleId] = rulePatches;
  });

  const missingIds = [...expectedIds].filter((id) => !seenIds.has(id));
  if (missingIds.length > 0) {
    const total = expectedIds.size;
    const tolerance = Math.max(2, Math.ceil(0.05 * total));

    for (const mid of [...missingIds].sort()) {
      warn(`${mid}: not found in model output, inserting null entry`);
      patches[mid] = {
        F3: { value: null, level: null, reasoning: "model_omitted" },
        F8: { value: null, level: null, reasoning: "model_omitted" },
      };
    }

    if (missingIds.length > tolerance) {
      const sortedMissing = [...missingIds].sort();
      fatal(
        `${missingIds.length} of ${total} rule IDs missing from model output ` +
          `(tolerance: ${tolerance}). Model may have truncated. ` +
          `Missing: ${sortedMissing.slice(0, 10).join(", ")}${missingIds.length > 10 ? "..." : ""}`
      );
    }
  }

  const unexpected = [...seenIds].filter((id) => !expectedIds.has(id));
  for (const uid of unexpected.sort()) {
    warn(`${uid}: in model output but not in scored_semi.json, ignoring`);
    delete patches[uid];
  }

  return patches;
}

function loadExpectedRuleIds(scoredSemiPath) {
  const data = JSON.parse(fs.readFileSync(scoredSemiPath, "utf-8"));

  const schemaVersion = data.schema_version;
  if (schemaVersion === null || schemaVersion === undefined) {
    warn("scored_semi.json has no schema_version field; assuming 0.1");
  } else if (schemaVersion !== "0.1") {
    fatal(
      `scored_semi.json schema_version is '${schemaVersion}', expected '0.1'. ` +
        `Pipeline version skew — regenerate scored_semi.json with the current scripts.`
    );
  }

  const rules = data.rules || [];
  return new Set(rules.filter((r) => r && Object.prototype.hasOwnProperty.call(r, "id")).map((r) => r.id));
}

function readStdin() {
  try {
    return fs.readFileSync(0, "utf-8");
  } catch {
    return "";
  }
}

function main() {
  const argv = process.argv.slice(2);
  if (argv.length < 1) {
    fatal(
      "Usage: parse_judgment.js <scored_semi.json> [--expected-ids R001,...] " +
        "[--input file] [--output file]"
    );
  }

  let inputPath = null;
  let outputPath = null;
  let expectedIdsOverride = null;
  const positional = [];
  let i = 0;
  while (i < argv.length) {
    if (argv[i] === "--input" && i + 1 < argv.length) {
      inputPath = argv[i + 1];
      i += 2;
    } else if (argv[i] === "--output" && i + 1 < argv.length) {
      outputPath = argv[i + 1];
      i += 2;
    } else if (argv[i] === "--expected-ids" && i + 1 < argv.length) {
      expectedIdsOverride = new Set(argv[i + 1].split(","));
      i += 2;
    } else {
      positional.push(argv[i]);
      i += 1;
    }
  }

  if (positional.length === 0) {
    fatal("Missing scored_semi.json argument");
  }
  const scoredSemiPath = positional[0];

  let allIds;
  try {
    allIds = loadExpectedRuleIds(scoredSemiPath);
  } catch (e) {
    fatal(`Cannot read scored_semi.json: ${e.message}`);
  }

  if (!allIds || allIds.size === 0) {
    fatal("scored_semi.json contains no rules");
  }

  const expectedIds = expectedIdsOverride || allIds;

  let rawInput;
  if (inputPath) {
    rawInput = fs.readFileSync(inputPath, "utf-8");
  } else {
    rawInput = readStdin();
  }
  if (!rawInput.trim()) {
    fatal("Empty input" + (inputPath ? ` from ${inputPath}` : " on stdin"));
  }

  const cleaned = stripFences(rawInput);

  let entries;
  try {
    entries = extractJsonArray(cleaned);
  } catch (e) {
    fatal(e.message);
  }

  const patches = buildPatches(entries, expectedIds);

  const output = {
    schema_version: "0.1",
    model_version: "unknown",
    patches,
  };
  if (outputPath) {
    fs.writeFileSync(outputPath, JSON.stringify(output, null, 2), "utf-8");
  } else {
    process.stdout.write(JSON.stringify(output, null, 2));
    process.stdout.write("\n");
  }
}

if (require.main === module) {
  main();
}

module.exports = {
  stripFences,
  extractJsonArray,
  resolveLevel,
  levelMidpoint,
  validateValueRange,
  truncateReasoning,
  validateFactor,
  validateEntry,
  buildPatches,
  loadExpectedRuleIds,
  main,
};
